### ISSUE TITLE
Number of calls to custom loss mismatches with total number of batch calls 

### ISSUE THREAD

**Issue by satyanarayan-rao** [keras-team/keras#14201]
Hi, 

I have a custom loss function defined where I am writing the loss value calculated every time to a file (snapshot of the code block is shown below): 
```
def custom_loss(self):
    @tf.function
    def loss(y_true,y_pred):
        ....
        # custom loss calcualtion code
        tf.print(K.mean(total_loss), output_stream=''.join(['file://',self.total_loss_file]))
        return total_loss
    return loss
```
I want to associate each value of `total_loss` written to file with its corresponding `epoch` and `batch_id`. But I am having a hard time doing that as I can't access these two values in this custom loss function. 

As an alternative, I wrote the following custom callback function and appended in the callbacks list:

```
class batch_and_epoch_logger(keras.callbacks.Callback):
    def __init__(self, batch_and_epoch_file = None):
        self.batch_and_epoch_fp = open(batch_and_epoch_file, "w")
        self.epoch_counter = 0
    def on_batch_end(self, batch, logs = None):
        batch_and_epoch = str(batch) + "_" + str(self.epoch_counter)
        self.batch_and_epoch_fp.write(str(batch) + "\t" + str(self.epoch_counter) + "\t" + batch_and_epoch  + "\n")
    def on_epoch_begin(self, epoch, logs = None):
        self.epoch_counter +=1
``` 
My expectation was that the number of lines in `batch_and_epoch_file` will be the same as the number of lines in `total_loss_file`. But it is not the case: with `20` epochs and a batch_size of `128` on a data of total rows  = `14531`, I am getting as expected `2280` lines (explained below) in `batch_and_epoch` file, but `2540` lines in `total_loss_file`. Here is my calculation: 

```
    Number of batches in one epoch = 14531/128 = 113.52 ~= 114
    Total number of default loss calls = 20*114 = 2280 
    total_loss_file overshoot by = 2540 - 2280 = 260 lines
```  

I am not really sure where to look for the potential issue. One possible reason I find is that: in a given batch, `custom_loss` is more than once. If that is the case, is it possible to trace what batches were with more than one call to the loss function? 

Any help will be greatly appreciated!

**Comment by satyanarayan-rao** [comment]
Figured it out, sorry for asking this stupid question: custom loss function is called for the validation data also. Total rows in the validation data: 1614, so 1614/128 = 12.6 ~=13, 20 epoch 13*20 = 260, exactly the overshoot. Hope this helps others!
